src/GraphQL/Instrumentation/GraphQLTelemetryProvider.cs

```diff
@@ -0,0 +1,183 @@
+
+    /// <inheritdoc/>
+    public virtual async Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next)
+    {
+        if (!_telemetryOptions.Filter(options))
+            return await next(options).ConfigureAwait(false);
+
+        // start the Activity, in fact Activity.Stop() will be called from within Activity.Dispose() at the end of using block 
+        using var activity = await StartActivityAsync(options).ConfigureAwait(false);
+
+        // do not record any telemetry if there are no listeners or it decided not to sample the current request
+        if (activity == null)
+            return await next(options).ConfigureAwait(false);
+
+        // record the requested operation name and optionally the GraphQL document
+        await SetInitialTagsAsync(activity, options).ConfigureAwait(false);
+
+        // record the operation type and the operation name (which may be specified within the
+        // document even if not specified in the request)
+        options.Listeners.Add(new TelemetryListener(this, activity, options));
+
+        // execute the request
+        ExecutionResult result;
+        try
+        {
+            result = await next(options).ConfigureAwait(false);
+        }
+        catch (Exception ex)
+        {
+            // Since DocumentExecuter catches and wraps most exceptions, this should only be thrown
+            // if the request was canceled or if ThrowOnUnhandledException is true. And since
+            // OperationCanceledExceptions are not a fault, we do not record them as such.
+            if (ex is not OperationCanceledException)
+            {
+#if NET6_0_OR_GREATER
+                activity.SetStatus(ActivityStatusCode.Error);
+#else
+                activity.SetTag("otel.status_code", "ERROR");
+#endif
+            }
+            _telemetryOptions.EnrichWithException(activity, ex);
+            throw;
+        }
+
+        // record the status
+        await SetResultTagsAsync(activity, options, result).ConfigureAwait(false);
+
+        // return the result
+        return result;
+    }
+
+    /// <summary>
+    /// Creates an <see cref="Activity"/> for the specified <see cref="ExecutionOptions"/> and starts it.
+    /// </summary>
+    protected virtual ValueTask<Activity?> StartActivityAsync(ExecutionOptions options)
+        => new(ActivitySource.StartActivity(ACTIVITY_OPERATION_NAME));
+
+    /// <summary>
+    /// Sets the <see cref="Activity"/> tags based on the specified <see cref="ExecutionOptions"/>.
+    /// The default implementation sets the <c>graphql.operation.name</c> and <c>graphql.document</c> tags.
+    /// </summary>
+    protected virtual Task SetInitialTagsAsync(Activity activity, ExecutionOptions options)
+    {
+        activity.SetTag("graphql.operation.name", options.OperationName);
+        if (_telemetryOptions.RecordDocument /* && activity.IsAllDataRequested */)
+            activity.SetTag("graphql.document", options.Query);
```

Comment:
```suggestion
        if (_telemetryOptions.RecordDocument /* && activity.IsAllDataRequested */)
        {
            var document = options.SanitizeDocument(activity, options);
            activity.SetTag("graphql.document", document);
        }
```

Reply:
Always set document or only when not null when returned from SanitizeDocument?

Reply:
`null` value will discarded by the sdk anyway

Reply:
Any SDK? Does spec say that null should be discarded?

Reply:
You are right. I'm always forgeting this should be SDK agnostic 😄
Disarding nulls was the behavior I observed, but [spect](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/README.md#attribute) says null is not valid value

> Attribute values of null are not valid and attempting to set a null value is undefined behavior.

Reply:
It seems in the [past](https://github.com/open-telemetry/opentelemetry-specification/issues/797), setting the `null` value was the way to remove the attribute. But this was [changed](https://github.com/open-telemetry/opentelemetry-specification/pull/992) to undefined behavior.

Reply:
According to the .NET API, setting `null` removes the attribute in .NET 5+.  See link here:

https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.activity.settag?view=net-5.0

Nevertheless, I have added if conditions to skip setting null tags.